### PR TITLE
Initialize snippets after solution load is complete

### PR DESCRIPTION
**Pull request by rchande**

### CODE REVIEW

**Comment by rchande** [comment]
Please review @dotnet/roslyn-ide @dpoeschl. The snippets integration tests pass locally for me, let me know if I missed something.

**Comment by rchande** [comment]
test vsi

**Comment by rchande** [comment]
retest vsi please

**Comment by tannergooding** [comment]
test windows_eta_open_prtest please
test windows_debug_unit32_prtest please
test windows_vsi_p1_open_prtest please
test windows_vsi_p2_open_prtest please

**Comment by rchande** [comment]
@dotnet-bot retest windows_vsi_p1_open_prtest please
// Previous failure: http://dotnet-ci.cloudapp.net/job/Private/job/dotnet_roslyn-internal/job/master/job/windows_vsi_p1_open_prtest/191/
// Retest reason:

**Comment by rchande** [comment]
Ping?

**Comment by rchande** [comment]
Ping

File: src/VisualStudio/Core/Def/Implementation/LanguageService/HACK_AbstractCreateServicesOnUiThread.cs [hunk 1]
```diff
@@ -52,6 +52,18 @@ void IWpfTextViewConnectionListener.SubjectBuffersDisconnected(IWpfTextView text
         /// </summary>
         public static void CreateServicesOnUIThread(IComponentModel componentModel, string languageName)
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Def/Implementation/LanguageService/HACK_AbstractCreateServicesOnUiThread.cs]: Why were these split up?  Why not do it all in the one function?

Comment (rchande) [inline src/VisualStudio/Core/Def/Implementation/LanguageService/HACK_AbstractCreateServicesOnUiThread.cs]: I wasn't sure if it was safe to do that, happy to move it if it is.

Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Def/Implementation/LanguageService/HACK_AbstractCreateServicesOnUiThread.cs]: Hrmm... i see the concern.  If you move it, then these types may be initialized from some BG thread which might break something.  Ok.  Completely ignore my original comment.  Your chnage LGTM.

File: src/VisualStudio/Core/Def/Implementation/Snippets/AbstractSnippetInfoService.cs [hunk 2]
```diff
@@ -113,35 +113,16 @@ private void PopulateSnippetCaches()
...
-            {
-                // Call the asynchronous IExpansionManager API from a background thread
-                Task.Factory.StartNew(async () => await PopulateSnippetCacheAsync(asyncExpansionManager).ConfigureAwait(false),
-                                CancellationToken.None,
-                                TaskCreationOptions.None,
-                                TaskScheduler.Default).CompletesAsyncOperation(token);
+            // In Dev14 Update2+ the platform always provides an IExpansion Manager
+            var asyncExpansionManager = (IExpansionManager)_expansionManager;
+            // Call the asynchronous IExpansionManager API from a background thread
+            Task.Factory.StartNew(async () => await PopulateSnippetCacheAsync(asyncExpansionManager).ConfigureAwait(false),
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Def/Implementation/Snippets/AbstractSnippetInfoService.cs]: you can remove the "async, await, and ConfigureAwait" from this expression entirely.

**Comment by CyrusNajmabadi** [comment]
LGTM with the tiny tweak i mentioned.

**Comment by crummel** [comment]
Build triggered. sha1 is merged.

**Comment by Pilchie** [comment]
:1:

**Comment by Pilchie** [comment]
retest this please

**Comment by Pilchie** [comment]
test vsi please

**Comment by rchande** [comment]
test vsi please

**Comment by rchande** [comment]
retest this please

**Comment by dpoeschl** [comment]
retest windows_eta_open_prtest please
retest windows_vsi_p1_open_prtest please

(the details links for these failed test suites are long gone, refreshing them...)

**Comment by dpoeschl** [comment]
retest windows_vsi_p1_open_prtest

**Comment by dpoeschl** [comment]
retest windows_vsi_p1_open_prtest please

**Comment by dpoeschl** [comment]
@rchande Is this part of perf work that needs to get in for `2.0 (Preview 5)`? I don't see a linked bug.

**Comment by rchande** [comment]
test vsi please

**Comment by rchande** [comment]
test vsi please

**Comment by rchande** [comment]
retest this please

**Comment by rchande** [comment]
retest windows_debug_vs-integration_prtest

File: src/VisualStudio/Core/Def/Implementation/Snippets/AbstractSnippetInfoService.cs [hunk 3]
```diff
@@ -113,32 +113,16 @@ private void PopulateSnippetCaches()
...
-            }
-            else
-            {
-                // Call the synchronous IVsExpansionManager API from the UI thread
-                Task.Factory.StartNew(() => PopulateSnippetCacheOnForeground(_expansionManager),
-                    CancellationToken.None,
-                    TaskCreationOptions.None,
-                    ForegroundTaskScheduler).CompletesAsyncOperation(token);                
-            }
+            // In Dev14 Update2+ the platform always provides an IExpansion Manager
```
Comment (dpoeschl) [inline src/VisualStudio/Core/Def/Implementation/Snippets/AbstractSnippetInfoService.cs]: @rchande Can we refer to it as "Visual Studio 2015 Update 2"?  "Dev14" seems too codewordy.

Comment (dpoeschl) [inline src/VisualStudio/Core/Def/Implementation/Snippets/AbstractSnippetInfoService.cs]: (10 months later 😃)

**Pull request merged**